### PR TITLE
[DeviceASAN][NFC] Code Restructure

### PR DESCRIPTION
**Pull request by AllanZyne**
LLVM: https://github.com/intel/llvm/pull/15843
code refactoring for implementing the msan layer

### CODE REVIEW

**Comment by AllanZyne** [comment]
Hi @oneapi-src/unified-runtime-maintain, could you please prioritize the review of this PR?
This PR has no functional change, and is essential for our further development!
Thank you very much!

**Review by zhaomaosu — APPROVED** [review]
lgtm

**Comment by AllanZyne** [comment]
I've made patch to fix this in LLVM part.

```
ninja: error: '/home/test-user/actions-runners/opencl-1/_work/unified-runtime/unified-runtime/ur-repo/source/loader/layers/sanitizer/asan_libdevice.hpp', needed by 'lib/libsycl-asan-amdgcn-amd-amdhsa.bc', missing and no known rule to make it
```

**Comment by AllanZyne** [comment]
Hi @pbalcer, could you please review and merge this PR before other sanitizer related PRs?
I've synced with other members.
Thank you very much!

File: source/loader/layers/sanitizer/sanitizer_common/sanitizer_stacktrace.cpp
Comment (yingcong-wu) [inline source/loader/layers/sanitizer/sanitizer_common/sanitizer_stacktrace.cpp:96]: I think we should keep this, maybe by creating a base class of option that have the common options available for all sanitizers, and in this case, the "debug" switch. But that can be addressed later.

Comment (AllanZyne) [reply source/loader/layers/sanitizer/sanitizer_common/sanitizer_stacktrace.cpp:96]: I agree. Besides, I started to prefer to create a single option object for all sanitizers. 
We can discuss this later.

**Comment by pbalcer** [comment]
> Hi @pbalcer, could you please review and merge this PR before other sanitizer related PRs? I've synced with other members. Thank you very much!

Will do, is the PR final? I see that you are still making changes.

**Comment by AllanZyne** [comment]
> > Hi @pbalcer, could you please review and merge this PR before other sanitizer related PRs? I've synced with other members. Thank you very much!
> 
> Will do, is the PR final? I see that you are still making changes.

Yes, it's final. I just pushed a commit to address a comment by Yingcong.

**Review by pbalcer — APPROVED** [review]
lgtm, just a few nits.

Not insisting, but I think you should drop the prefixes on files if they are not necessary, so "common" instead of "sanitizer_common", "allocator.hpp" instead of "sanitizer_allocator.hpp" and so on.

**Comment by pbalcer** [comment]
I'll merge this once you get an approval on your intel/llvm PR (you are missing llvm-reviewers-runtime, I'll approve on behalf of unified-runtime-reviewers).

**Comment by AllanZyne** [comment]
> lgtm, just a few nits.
> 
> Not insisting, but I think you should drop the prefixes on files if they are not necessary, so "common" instead of "sanitizer_common", "allocator.hpp" instead of "sanitizer_allocator.hpp" and so on.

For file naming style, I completely use the same way in compiler_rt (https://github.com/intel/llvm/tree/sycl/compiler-rt/lib/sanitizer_common) for convenience.

**Comment by pbalcer** [comment]
> > lgtm, just a few nits.
> > Not insisting, but I think you should drop the prefixes on files if they are not necessary, so "common" instead of "sanitizer_common", "allocator.hpp" instead of "sanitizer_allocator.hpp" and so on.
> 
> For file naming style, I completely use the same way in compiler_rt (https://github.com/intel/llvm/tree/sycl/compiler-rt/lib/sanitizer_common) for convenience.

Ok, makes sense, it's better to be consistent. Disregard that then.

File: source/loader/layers/sanitizer/asan/asan_interceptor.cpp
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan/asan_interceptor.cpp:927]: > But can't we just leave the AsanInterceptor inside of the layer object?

Yes, I agree that using global variable is not a good idea. 
We are going to develop each sanitizer as independent as possible at the beginning, and when these features are stable, we will try to extract their common characteristics, and create some base classes then.
Now I want the outside world (layer context) don't know how each sanitizer is implemented, and just give it some simple interfaces.

**Pull request merged**
